### PR TITLE
Split Buffer into processing and update buffers

### PR DESCRIPTION
**Pull request by ervteng**
This is a series of PRs that intend to move the agent processing logic (add_experiences and process_experiences) out of the trainer and into a separate class. The plan is to do so in steps:

- Split the processing buffers (keeping track of agent trajectories and assembling trajectories) and update buffer (complete trajectories to be used for training) within the Trainer (this PR)
- Move the processing buffer and add/process experiences into a separate, outside class
- Change the data type of the update buffer to be a Trajectory
- Place and read Trajectories from queues, add subscription mechanism for both AgentProcessor and Trainers

### CODE REVIEW

File: ml-agents/MANIFEST.in [hunk 1]
```diff
@@ -0,0 +1 @@
+include ../VERSION
```
Comment (chriselion) [inline ml-agents/MANIFEST.in]: remove?

File: ml-agents/mlagents/trainers/buffer.py [hunk 2]
```diff
@@ -12,273 +12,253 @@ class BufferException(UnityException):
...
-                Resets the AgentBufferField
-                """
-                self[:] = []
-
         def __init__(self):
-            self.last_brain_info = None
-            self.last_take_action_outputs = None
-            super(Buffer.AgentBuffer, self).__init__()
+            self.padding_value = 0
+            super(AgentBuffer.AgentBufferField, self).__init__()
```
Comment (chriselion) [inline ml-agents/mlagents/trainers/buffer.py]:
```suggestion
            super().__init__()
```

File: ml-agents/mlagents/trainers/buffer.py [hunk 3]
```diff
@@ -12,273 +12,253 @@ class BufferException(UnityException):
...
+            type, so that it can be set on append (e.g. action_masks should
+            be padded with 1.)
+            :param element: The element to append to the list.
+            :param padding_value: The value used to pad when get_batch is called.
             """
-            for k in self.keys():
-                self[k].reset_field()
-            self.last_brain_info = None
-            self.last_take_action_outputs = None
+            super(AgentBuffer.AgentBufferField, self).append(element)
```
Comment (chriselion) [inline ml-agents/mlagents/trainers/buffer.py]:
```suggestion
            super().append(element)
```

File: ml-agents/mlagents/trainers/buffer.py [hunk 4]
```diff
@@ -12,273 +12,253 @@ class BufferException(UnityException):
...
-                    # extend() will convert the numpy array's first dimension into list
-                    self[key].extend(read_file[key][()])
+            self[:] = []
 
     def __init__(self):
-        self.update_buffer = self.AgentBuffer()
-        super(Buffer, self).__init__()
+        self.last_brain_info = None
+        self.last_take_action_outputs = None
+        super(AgentBuffer, self).__init__()
```
Comment (chriselion) [inline ml-agents/mlagents/trainers/buffer.py]:
```suggestion
        super().__init__()
```

File: ml-agents/mlagents/trainers/buffer.py [hunk 5]
```diff
@@ -12,273 +12,253 @@ class BufferException(UnityException):
...
+            self[k].reset_field()
+        self.last_brain_info = None
+        self.last_take_action_outputs = None
 
     def __getitem__(self, key):
         if key not in self.keys():
-            self[key] = self.AgentBuffer()
-        return super(Buffer, self).__getitem__(key)
+            self[key] = self.AgentBufferField()
+        return super(AgentBuffer, self).__getitem__(key)
```
Comment (chriselion) [inline ml-agents/mlagents/trainers/buffer.py]:
```suggestion
        return super().__getitem__(key)
```

File: ml-agents/mlagents/trainers/buffer.py [hunk 6]
```diff
@@ -12,273 +12,253 @@ class BufferException(UnityException):
...
         """
-        self.update_buffer.reset_agent()
+        with h5py.File(file_object) as read_file:
+            for key in list(read_file.keys()):
+                self[key] = AgentBuffer.AgentBufferField()
+                # extend() will convert the numpy array's first dimension into list
+                self[key].extend(read_file[key][()])
 
-    def truncate_update_buffer(self, max_length, sequence_length=1):
+    def truncate(self, max_length, sequence_length=1):
```
Comment (chriselion) [inline ml-agents/mlagents/trainers/buffer.py]: nit: add type hints

File: ml-agents/mlagents/trainers/buffer.py [hunk 7]
```diff
@@ -289,7 +269,12 @@ def reset_local_buffers(self):
             self[k].reset_agent()
 
     def append_update_buffer(
-        self, agent_id, key_list=None, batch_size=None, training_length=None
+        self,
```
Comment (chriselion) [inline ml-agents/mlagents/trainers/buffer.py]: nit: add type hints

Comment (chriselion) [inline ml-agents/mlagents/trainers/buffer.py]: This name feels more misleading now - it sounds like you're taking the `update_buffer` and appending it to `self`, but it's actually the opposite.

How about either renaming to `append_to_update_buffer`, or make `append_processing_buffer` a method of `AgentBuffer`? The latter is a bigger change but I think it's a bit cleaner.

Comment (chriselion) [reply ml-agents/mlagents/trainers/buffer.py]: (if nothing else, `append_to_update_buffer` is kinda close to `append_all_agent_batch_to_update_buffer`)

File: ml-agents/mlagents/trainers/demo_loader.py [hunk 8]
```diff
@@ -39,33 +40,35 @@ def make_demo_buffer(
...
-            demo_buffer.append_update_buffer(
-                0, batch_size=None, training_length=sequence_length
+            demo_process_buffer.append_update_buffer(
+                demo_buffer, 0, batch_size=None, training_length=sequence_length
             )
-            demo_buffer.reset_local_buffers()
-    demo_buffer.append_update_buffer(
-        0, batch_size=None, training_length=sequence_length
+            demo_process_buffer.reset_local_buffers()
+    demo_process_buffer.append_update_buffer(
```
Comment (chriselion) [inline ml-agents/mlagents/trainers/demo_loader.py]: For example, I think this would be clearer with `append_update_buffer` renamed (or making it `demo_buffer.append_processing_buffer(...)`)

File: ml-agents/mlagents/trainers/buffer.py (hunk 7 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/buffer.py]: `append_to_update_buffer` is pretty close. Eventually this method will be replaced by `add_trajectory_to_queue` so it probably shouldn't live in the low-level `AgentBuffer`.

Comment (ervteng) [reply ml-agents/mlagents/trainers/buffer.py]: `Done()`

Comment (chriselion) [reply ml-agents/mlagents/trainers/buffer.py]: SG. 

When I meant "append_to_update_buffer" is close to "append_all_agent_batch_to_update_buffer", I meant it in a good way (more consistent), not that they'd be mistaken for each other.

Comment (ervteng) [reply ml-agents/mlagents/trainers/buffer.py]: Oh yeah, that makes sense 😄 Suppose it's technically `append_agent_trajectory_to_update_buffer`

**Comment by vincentpierre** [comment]
Are there any performance changes (I mean in training speed) ?

**Comment by ervteng** [comment]
> Are there any performance changes (I mean in training speed) ?

Not as far as I can tell (at least on 3DBall). There shouldn't be, given that it's the same # of objects, just that the update buffer is outside the class of the AgentBuffers.

File: ml-agents/mlagents/trainers/bc/trainer.py [hunk 9]
```diff
@@ -9,7 +9,7 @@
 from mlagents.envs.brain import BrainInfo
 from mlagents.envs.action_info import ActionInfoOutputs
 from mlagents.trainers.bc.policy import BCPolicy
-from mlagents.trainers.buffer import Buffer
+from mlagents.trainers.buffer import AgentBuffer, AgentProcessorBuffer
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/bc/trainer.py]: Maybe the names are not descriptive enough about the difference between the two.

File: ml-agents/mlagents/trainers/buffer.py [hunk 10]
```diff
@@ -12,285 +13,277 @@ class BufferException(UnityException):
...
         if current_length > max_length:
-            for _key in self.update_buffer.keys():
-                self.update_buffer[_key] = self.update_buffer[_key][
-                    current_length - max_length :
-                ]
+            for _key in self.keys():
+                self[_key] = self[_key][current_length - max_length :]
+
+
+class AgentProcessorBuffer(dict):
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/buffer.py]: Can we use separate files? I think it would reduce some confusion.

File: ml-agents/mlagents/trainers/buffer.py [hunk 11]
```diff
@@ -12,285 +13,277 @@ class BufferException(UnityException):
...
+    """
+    AgentProcessorBuffer contains a dictionary of AgentBuffer. The AgentBuffers are indexed by agent_id.
+    Buffer also contains an update_buffer that corresponds to the buffer used when updating the model.
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/buffer.py]: Reference to `Buffer` here. I think you meant "AgentProcessorBuffer". Also, does it really have an update_buffer still?

File: ml-agents/mlagents/trainers/bc/trainer.py [hunk 12]
```diff
@@ -112,26 +112,24 @@ def is_ready_update(self):
         Returns whether or not the trainer has enough elements to run update model
         :return: A boolean corresponding to whether or not update_model() can be run
         """
-        return (
-            len(self.demonstration_buffer.update_buffer["actions"]) > self.n_sequences
-        )
+        return len(self.demonstration_buffer["actions"]) > self.n_sequences
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/bc/trainer.py]: Maybe the AgentBuffer should have a `get_len()` method, we seem to be calling `len(self.demonstration_buffer["actions"])` a lot.

Comment (chriselion) [reply ml-agents/mlagents/trainers/bc/trainer.py]: Or `get_num_demonstrations()`? Anything with `len` in it could be confusing, and overriding `__len__()` is likely to cause bugs.

Comment (ervteng) [reply ml-agents/mlagents/trainers/bc/trainer.py]: Agreed. In the AgentBuffer sample method I use `len(next(iter(self.values())))` to get the length of the first array. I'll wrap this into a public function

File: ml-agents/mlagents/trainers/bc/trainer.py (hunk 9 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/bc/trainer.py]: What about calling it the `ProcessingBuffer`? 

In the next PR I'm actually removing the AgentProcessorBuffer entirely and baking the logic into the AgentProcessor itself.

File: ml-agents/mlagents/trainers/bc/trainer.py (hunk 12 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/bc/trainer.py]: Added a property called `num_experiences` in line with the future AgentExperiences

File: ml-agents/mlagents/trainers/buffer.py (hunk 11 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/buffer.py]: deleted this one

File: ml-agents/mlagents/trainers/buffer.py (hunk 10 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/buffer.py]: Split into separate files -> `agent_processor.py` contains the processing buffer

File: ml-agents/mlagents/trainers/bc/trainer.py (hunk 9 above)
Comment (vincentpierre) [reply ml-agents/mlagents/trainers/bc/trainer.py]: Processing is a little vague. I would say maybe SingleAgentExperienceBuffer, MultiAgentExperienceBuffer ?

File: ml-agents/mlagents/trainers/agent_processor.py [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+        )
+
+    def __getitem__(self, key):
+        if key not in self.keys():
+            self[key] = AgentBuffer()
+        return super().__getitem__(key)
+
+    def reset_local_buffers(self) -> None:
+        """
+        Resets all the local local_buffers
```
Comment (chriselion) [inline ml-agents/mlagents/trainers/agent_processor.py]:
```suggestion
        Resets all the local buffers
```

File: ml-agents/mlagents/trainers/agent_processor.py [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+        """
+        agent_ids = list(self.keys())
```
Comment (chriselion) [inline ml-agents/mlagents/trainers/agent_processor.py]: rewrite as 
``` python
for buf in self.values():
    buf.reset_agent()
```
(or something like that) ?  Forming the list is unnecessary (allocates memory) and you're doing key lookups that you don't really need.

File: ml-agents/mlagents/trainers/buffer.py [hunk 15]
```diff
@@ -12,321 +13,260 @@ class BufferException(UnityException):
...
+    ) -> "AgentBuffer":
+        """
+        Creates a mini-batch from a random start and end.
+        :param batch_size: number of elements to withdraw.
+        :param sequence_length: Length of sequences to sample.
+            Number of sequences to sample will be batch_size/sequence_length.
+        """
+        num_seq_to_sample = batch_size // sequence_length
+        mini_batch = AgentBuffer()
+        buff_len = len(next(iter(self.values())))
```
Comment (chriselion) [inline ml-agents/mlagents/trainers/buffer.py]: nitL use self.num_experiences here?

File: ml-agents/mlagents/trainers/buffer.py [hunk 16]
```diff
@@ -12,321 +13,260 @@ class BufferException(UnityException):
...
-                "The length of the fields {0} for agent {1} where not of same length".format(
-                    key_list, agent_id
-                )
-            )
-        for field_key in key_list:
-            self.update_buffer[field_key].extend(
-                self[agent_id][field_key].get_batch(
-                    batch_size=batch_size, training_length=training_length
-                )
+                "Unable to shuffle if the fields are not of same length"
```
Comment (chriselion) [inline ml-agents/mlagents/trainers/buffer.py]: This error message doesn't necessarily make sense here (you might not be shuffling at the time).

File: ml-agents/mlagents/trainers/agent_processor.py (hunk 14 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/agent_processor.py]: Done

File: ml-agents/mlagents/trainers/sac/trainer.py [hunk 17]
```diff
@@ -93,7 +93,7 @@ def __init__(
                 )
             LOGGER.debug(
                 "Loaded update buffer with {} sequences".format(
-                    len(self.training_buffer.update_buffer["actions"])
+                    len(self.update_buffer["actions"])
```
Comment (chriselion) [inline ml-agents/mlagents/trainers/sac/trainer.py]: `self.update.buffer.num_experiences`?

File: ml-agents/mlagents/trainers/sac/trainer.py [hunk 18]
```diff
@@ -127,10 +127,10 @@ def load_replay_buffer(self) -> None:
         filename = os.path.join(self.policy.model_path, "last_replay_buffer.hdf5")
         LOGGER.info("Loading Experience Replay Buffer from {}".format(filename))
         with open(filename, "rb+") as file_object:
-            self.training_buffer.update_buffer.load_from_file(file_object)
+            self.update_buffer.load_from_file(file_object)
         LOGGER.info(
             "Experience replay buffer has {} experiences.".format(
-                len(self.training_buffer.update_buffer["actions"])
+                len(self.update_buffer["actions"])
```
Comment (chriselion) [inline ml-agents/mlagents/trainers/sac/trainer.py]: self.update_buffer.num_experiences?

File: ml-agents/mlagents/trainers/sac/trainer.py [hunk 19]
```diff
@@ -283,11 +285,8 @@ def update_sac_policy(self) -> None:
 
         # Truncate update buffer if neccessary. Truncate more than we need to to avoid truncating
         # a large buffer at each update.
-        if (
-            len(self.training_buffer.update_buffer["actions"])
-            > self.trainer_parameters["buffer_size"]
-        ):
-            self.training_buffer.truncate_update_buffer(
+        if len(self.update_buffer["actions"]) > self.trainer_parameters["buffer_size"]:
```
Comment (chriselion) [inline ml-agents/mlagents/trainers/sac/trainer.py]: self.update_buffer.num_experiences?

File: ml-agents/mlagents/trainers/buffer.py (hunk 16 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/buffer.py]: Oh no, I left some test code in... checking the lengths of the fields every time this was called made things significantly slower, so I removed it.

File: ml-agents/mlagents/trainers/buffer.py (hunk 15 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/buffer.py]: I think I finally got all of these calls out
